### PR TITLE
Handle operating system shutdown by overriding OnShutdown()

### PR DESCRIPTION
**Pull request by peval27**
Summary of the changes
 - overridden ServiceBase.OnShutdown() to get notified on OS shutdown

Addresses #2699

### CODE REVIEW

**Comment by anurse** [comment]
I was about to ask "can you add a test" but then I realized that would be **very** tricky ;).

**Comment by peval27** [comment]
Yeah I don't see too much existing unit testing here

**Comment by peval27** [comment]
BTW I saw you added milestone 5.0.0. what's the ETA? The released one is 3, so I assume quite a long time here. Wouldn't be possible to squash his big fix in a maintenance 3.0.x

**Comment by anurse** [comment]
> Yeah I don't see too much existing unit testing here

Well, and testing something that requires shutting down the machine doesn't really lend itself to an automated test ;).

The full roadmap can be found here: https://github.com/dotnet/core/blob/master/roadmap.md

> BTW I saw you added milestone 5.0.0. what's the ETA? The released one is 3, so I assume quite a long time here. Wouldn't be possible to squash his big fix in a maintenance 3.0.x

5.0 is the next release after 3.1 (which is locked down for release in December), we're skipping 4.x to avoid conflicts with .NET Framework. The ETA is November 2020.

We don't generally add functionality in a maintenance release. We never provided support for the `OnShutdown` event in the past so this isn't really a regression. Can you provide any details on what scenarios are blocked in your environment? Given that this can be worked around by registering your own version of `WindowsServiceLifetime` in the container (based on this fixed code), I think this wouldn't meet the servicing bar.

File: src/Hosting/WindowsServices/src/WindowsServiceLifetime.cs [hunk 1]
```diff
@@ -90,6 +91,12 @@ protected override void OnStop()
             _delayStop.Wait(_hostOptions.ShutdownTimeout);
             base.OnStop();
         }
+        
+        protected override void OnShutdown()
+        {
+             ApplicationLifetime.StopApplication();
```
Comment (Tratcher) [inline src/Hosting/WindowsServices/src/WindowsServiceLifetime.cs]: If OnShutdown blocks will that delay shutdown? Does the application have time to actually shut down before the OS kills it? Compare to OnStop.

Comment (peval27) [reply src/Hosting/WindowsServices/src/WindowsServiceLifetime.cs]: Good point. will fix

Comment (peval27) [reply src/Hosting/WindowsServices/src/WindowsServiceLifetime.cs]: good point, will fix it.

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**